x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportCloseJobAction.java

```diff
@@ -249,21 +250,97 @@ void stopDatafeedsIfNecessary(OpenAndClosingIds jobIds, boolean isForce, TimeVal
+                            // A datafeed with an end time will gracefully close its job when it stops even if it was force stopped.
+                            // If we didn't do anything about this then it would turn requests to force close jobs into normal close
+                            // requests for those datafeeds, which is undesirable - the caller specifically asked for the job to be
+                            // closed forcefully, skipping the final state persistence to save time.  Therefore, before stopping the
+                            // datafeeds in this case we isolate them.  An isolated datafeed will NOT close its associated job under
+                            // any circumstances.  The downside of doing this is that if the stop datafeeds call fails then this API
+                            // will not have closed any jobs, but will have isolated one or more datafeeds, so the failure will have
+                            // an unexpected side effect.  Hopefully the project to combine jobs and datafeeds will be able to improve
+                            // on this.
+                            isolateDatafeeds(jobIds.openJobIds, runningDatafeedIds, ActionListener.wrap(
+                                r -> stopDatafeeds(runningDatafeedIds, true, timeout, listener),
+                                listener::onFailure
+                            ));
+                        } else {
+                            stopDatafeeds(runningDatafeedIds, false, timeout, listener);
+                        }
                     }
                 },
                 listener::onFailure
         ));
     }
 
+    private void stopDatafeeds(List<String> runningDatafeedIds, boolean isForce, TimeValue timeout,
+                               ActionListener<Boolean> listener) {
+        StopDatafeedAction.Request request = new StopDatafeedAction.Request(String.join(",", runningDatafeedIds));
+        request.setForce(isForce);
+        request.setStopTimeout(timeout);
+        ClientHelper.executeAsyncWithOrigin(
+            client,
+            ClientHelper.ML_ORIGIN,
+            StopDatafeedAction.INSTANCE,
+            request,
+            ActionListener.wrap(
+                r -> listener.onResponse(r.isStopped()),
+                e -> listener.onFailure(ExceptionsHelper.conflictStatusException(
+                    "failed to close jobs as one or more had started datafeeds that could not be stopped: " +
+                        "started datafeeds [{}], error stopping them [{}]", e, request.getDatafeedId(), e.getMessage())
+                )
+            )
+        );
+    }
+
+    void isolateDatafeeds(List<String> openJobs, List<String> runningDatafeedIds, ActionListener<Void> listener) {
+
+        final int numberOfDatafeeds = runningDatafeedIds.size();
+        assert numberOfDatafeeds > 0;
+        final AtomicInteger counter = new AtomicInteger();
+
+        for (String runningDatafeedId : runningDatafeedIds) {
```

Comment:
Seems like a good use of `GroupedActionListener` though it may lose the `info` logging (if that is super important or not).
```
ActionListener<?> groupedActionListener = new GroupedActionListener<?>(ActionListenener.wrap(r -> ...., e->...);
for (String runningDatafeedId : runningDatafeedIds) {
...
ClientHelper.executeAsyncWithOrigin(
                client,
                ClientHelper.ML_ORIGIN,
                IsolateDatafeedAction.INSTANCE,
                request,
                groupedActionListener);
}
```

Reply:
> though it may lose the info logging (if that is super important or not).

It doesn't completely lose it.  It means there is only one message instead of many in the case where multiple datafeeds cannot be isolated, and the datafeed IDs are only mentioned in the exceptions then, and not in the main message.  Since this will be such a super-rare occurrence and it will be us interpreting the logs I don't think it really matters.